### PR TITLE
Promise-aware config reading and support for `plugins` configuration

### PR DESCRIPTION
**Pull request by lukemelia**
The first commit in this PR was before I realized that ConfigurationReader was not used by the pipeline, so it's good to have it merge, but not relevant to pipeline task operation.

### CODE REVIEW

File: lib/tasks/pipeline.js [hunk 1]
```diff
@@ -38,58 +36,83 @@ module.exports = Task.extend({
...
     var isValidDeployPlugin = this._isValidDeployPlugin(addon);
+    if (!isValidDeployPlugin) { return; }
 
-    if (isValidDeployPlugin) {
-      var pluginNameRegex = /^(ember\-cli\-deploy\-)(.*)$/;
-      var name            = addon.name.match(pluginNameRegex)[2];
-      var pluginHooks     = addon.createDeployPlugin({name: name});
+    var pluginNameRegex = /^(ember\-cli\-deploy\-)(.*)$/;
+    var name            = addon.name.match(pluginNameRegex)[2];
+    if (configuredPluginList && configuredPluginList.indexOf(name) === -1) {
```
Comment (achambers) [inline lib/tasks/pipeline.js]: @lukemelia , Am I right in saying that this currently expects the `configuredPluginList` to be an array of the actual plugin in names, eg `['build', 'tag', 's3', 'redis']`.....as opposed to the goal of the advanced config being `['build', 'tag', 's3:index', 's3:assets']` ? I'm guessing this is a step in that direction yeh?

Comment (lukemelia) [inline lib/tasks/pipeline.js]: Yes, it is just plugin names, and does not yet support naming the plugin instances yet.

Comment (achambers) [inline lib/tasks/pipeline.js]: Cool.  Just checking that my understanding was correct.  Nice work

**Comment by achambers** [comment]
Thanks @lukemelia . Love ya work ;)

**Pull request merged**